### PR TITLE
Conversion to and from Helmholtz

### LINKED ISSUE #38: Helmholtz pitch notation
https://en.wikipedia.org/wiki/Helmholtz_pitch_notation

Should maybe support converting pitches to (+ from?) strings of this format, as an alternative to SPN.

### PR DESCRIPTION
**Pull request by mccalluc**
- Fix #38

Hope this is on the right track, but please comment if not.

In `Pitch.from_name`, `_class_name` and `_octave` are being set on the new object: Is this something we should do in every case?

The logic overlaps with `_octave_from_abc_parts`: Easiest just to leave it as-is, but another possibility would be to use the new `from_helmholtz` here.

### CODE REVIEW

File: pyabc2/pitch.py [hunk 1]
```diff
@@ -498,6 +498,28 @@ def _repr_html_(self):
         cn = self.to_pitch_class()._repr_html_()
         return f"{cn}<sub>{self.octave}</sub>"
 
+    def helmholtz(self) -> str:
```
Comment (zmoon) [inline pyabc2/pitch.py:501]: I think this could be a property

File: pyabc2/pitch.py [hunk 2]
```diff
@@ -498,6 +498,28 @@ def _repr_html_(self):
...
+        """Note name in Helmholz pitch notation, e.g. "C," and "c'"."""
```
Comment (zmoon) [inline pyabc2/pitch.py:502]:
```suggestion
        """Pitch name in Helmholtz pitch notation, e.g. ``C,`` and ``c'`` (ASCII)."""
```
I know it isn't very consistent currently, but I try to distinguish between note (pitch + duration) and pitch (no duration), since they each have their own class.

File: pyabc2/pitch.py [hunk 3]
```diff
@@ -498,6 +498,28 @@ def _repr_html_(self):
...
+        if little_c_octave >= 0:
+            return self.class_name.lower() + "'" * little_c_octave
+        big_c_octave = -self.octave + 2
+        return self.class_name + "," * big_c_octave
+
+    @classmethod
+    def from_helmholtz(cls, helmholtz_name: str) -> "Pitch":
+        helmholtz_name = helmholtz_name.strip()
+        is_upper = helmholtz_name[0].isupper()
+        helmoltz_re = r"([^,]+)(,*)" if is_upper else r"([^']+)('*)"
```
Comment (zmoon) [inline pyabc2/pitch.py:513]: It might better to be more explicit here, use `[A-G]` and `[a-g]` etc. As it is now, `Pitch.from_helmholtz("c,")` passes the matching here but errors in `.from_name` in a way that is a bit confusing, for example.

File: pyabc2/pitch.py [hunk 4]
```diff
@@ -577,6 +599,7 @@ def from_name(cls, name: str) -> "Pitch":
         class_value = pitch_class_value(class_name)
 
         p = cls.from_class_value(class_value, octave)
+        # TODO: Why are these attributes being set? Should other from_* methods do the same?
```
Comment (zmoon) [inline pyabc2/pitch.py:602]:
```suggestion
```
That is a great question. I realize it's a bit convoluted. It's a hack I put in to allow preserving "non-standard" pitch class name input like Cb etc., which also affects the value since the octave is set by the natural note name. For example, `Cb,` and `B,` are an octave apart.

Since you called `.from_class_name`, which calls `.from_name`, this already works.

Maybe you could add a test case or two for this.

File: tests/test_note.py [hunk 5]
```diff
@@ -556,3 +556,53 @@ def test_note_to_from_nonimpl(meth):
...
+        ("C4", "c'"),
+        ("C5", "c''"),
+        ("B3", "b"),
+        ("B#3", "b#"),
+        ("D3", "d"),
+    ],
+)
+def test_helmholtz_from_pitch(scientific, expected_helmholtz):
+    pitch = Pitch.from_name(scientific)
+    assert pitch.helmholtz() == expected_helmholtz
```
Comment (zmoon) [inline tests/test_note.py:576]: This is probably sufficient, but could also check that they have the same `.name`.

File: pyabc2/pitch.py [hunk 6]
```diff
@@ -498,6 +498,28 @@ def _repr_html_(self):
...
+        return self.class_name + "," * big_c_octave
+
+    @classmethod
+    def from_helmholtz(cls, helmholtz_name: str) -> "Pitch":
+        helmholtz_name = helmholtz_name.strip()
+        is_upper = helmholtz_name[0].isupper()
+        helmoltz_re = r"([^,]+)(,*)" if is_upper else r"([^']+)('*)"
+        m = re.fullmatch(helmoltz_re, helmholtz_name)
+        if m is None:
+            raise ValueError(f"invalid helmholtz pitch name '{helmholtz_name}'")
```
Comment (zmoon) [inline pyabc2/pitch.py:516]:
```suggestion
            raise ValueError(f"invalid Helmholtz pitch name '{helmholtz_name}'")
```

**Review by zmoon — COMMENTED** [review]
This is great, thanks.

> The logic overlaps with `_octave_from_abc_parts`: Easiest just to leave it as-is

👍

File: pyabc2/pitch.py (hunk 3 above)
Comment (zmoon) [reply pyabc2/pitch.py:513]: Also might as well add a short docstring if you don't mind.

**Comment by mccalluc** [comment]
That all makes sense. I had held off on a tighter regex because it felt like the concerns were overlapping, but you're right, having a clear exception thrown hear is better than getting a cryptic one from deeper in the code.

I'll make this a draft, tweak, push, and re-request review. Thanks!

File: pyabc2/pitch.py (hunk 1 above)
Comment (mccalluc) [reply pyabc2/pitch.py:501]:
```suggestion
    @property
    def helmholtz(self) -> str:
```

File: tests/test_note.py (hunk 5 above)
Comment (mccalluc) [reply tests/test_note.py:576]: I don't think I understand the exact request here. `expected_helmholtz` is just a string, so it wouldn't have a `.name` property. I could check the pitch class? I'll also add a c-flat example, since it's another good edge case...

File: tests/test_note.py [hunk 7]
```diff
@@ -26,7 +26,6 @@
         ("C", 0),
         ("C#", 1),
         ("Db", 1),
-        ("C###", 3),
```
Comment (mccalluc) [inline tests/test_note.py:29]: We now use the same regex on different execution paths, so triple accidentals will no longer parse.

File: tests/test_note.py [hunk 8]
```diff
@@ -556,3 +554,49 @@ def test_note_to_from_nonimpl(meth):
...
+class TestHelholtz:
+    def test_helmholtz_from_pitch(self, scientific, helmholtz):
+        pitch = Pitch.from_name(scientific)
+        assert pitch.helmholtz == helmholtz
+
+    def test_pitch_from_helmholtz(self, scientific, helmholtz):
+        pitch = Pitch.from_helmholtz(helmholtz)
+        assert pitch.name == scientific
+
+    def test_helmholtz_name_and_octave(self, scientific, helmholtz):
```
Comment (mccalluc) [inline tests/test_note.py:582]: Not sure if this is what you meant by checking the name...

Comment (zmoon) [reply tests/test_note.py:582]: Yes that is what I was thinking, thanks. Sorry it wasn't super clear. Could check `.name` is same too because why not.

File: tests/test_note.py (hunk 7 above)
Comment (zmoon) [reply tests/test_note.py:29]: That's fine, working for arbitrary combos of `#b=` in `pitch_class_value` was only for fun anyway, not very useful. Maybe I'll bring it back as an option later.

File: pyabc2/pitch.py [hunk 9]
```diff
@@ -58,7 +58,9 @@ def _gen_pitch_values() -> Dict[str, int]:
 The more common accidentals are used.
 """
 
-_S_RE_PITCH_CLASS = r"[A-G][\#b\=]*"
+_S_RE_ASCII_ACIDENTALS = r"(?:##|bb|b|#|=)"
```
Comment (zmoon) [inline pyabc2/pitch.py:61]:
```suggestion
_S_RE_ASCII_ACCIDENTALS = r"(?:##|bb|b|#|=)"
```

File: tests/test_note.py [hunk 10]
```diff
@@ -556,3 +554,50 @@ def test_note_to_from_nonimpl(meth):
...
+        ("C2", "C"),
+        ("C3", "c"),
+        ("C4", "c'"),
+        ("C5", "c''"),
+        ("B3", "b"),
+        ("B#3", "b#"),
+        ("Cb2", "Cb"),
+    ],
+)
+class TestHelholtz:
```
Comment (zmoon) [inline tests/test_note.py:573]:
```suggestion
class TestHelmholtz:
```

**Review by zmoon — CHANGES_REQUESTED** [review]
Thanks, I like what you did with the regexes.

**Review by zmoon — APPROVED** [review]
